### PR TITLE
Sdw/endpoint aggregation

### PR DESCRIPTION
**Pull request by plbossart**
This PR shows how we can expand the ACPI _ADR based information to clarify how endpoints can be logically grouped, e.g. handle 2 speaker amplifiers on separate links as a single logical output.

@RanderWang this is an alternative to your initial proposal, I believe it would simplify the decision to use multiple cpu-dais for  link, essentially if you find a device with an aggregated endpoint you need more than one cpu-dai per dailink.

@bardliao please double check that the table values are consistent, I *hope* I didn't break too many things on a Friday afternoon :-)

### CODE REVIEW

File: sound/soc/intel/common/soc-acpi-intel-cml-match.c [hunk 1]
```diff
@@ -59,42 +59,96 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_cml_machines[] = {
 };
 EXPORT_SYMBOL_GPL(snd_soc_acpi_intel_cml_machines);
 
-static const u64 rt711_0_adr[] = {
-	0x000010025D071100
+static const struct snd_soc_acpi_endpoint single_endpoint = {
+	0,
+	0,
+	0,
+	0,
```
Comment (lyakh) [inline sound/soc/intel/common/soc-acpi-intel-cml-match.c]: why not use named initialisation like
```
static const struct snd_soc_acpi_endpoint single_endpoint = {
	.num = 0,
	.aggregated = 0,
	.group_position = 0,
	.group_id = 0,
};
```

File: sound/soc/intel/common/soc-acpi-intel-cml-match.c [hunk 2]
```diff
@@ -59,42 +59,96 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_cml_machines[] = {
...
 };
 
-static const u64 rt1308_1_adr[] = {
-	0x000110025D130800
+static const struct snd_soc_acpi_endpoint spk_l_endpoint = {
+	0,
+	1, /* aggregated */
+	0, /* group_position */
+	1, /* group_id */
```
Comment (lyakh) [inline sound/soc/intel/common/soc-acpi-intel-cml-match.c]: ditto, then comments would become redundant. Same below

File: sound/soc/intel/common/soc-acpi-intel-cml-match.c [hunk 3]
```diff
@@ -59,42 +59,96 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_cml_machines[] = {
...
 };
 
-static const u64 rt715_3_adr[] = {
-	0x000310025D071500
+static const struct snd_soc_acpi_adr_device rt711_0_adr[] = {
+	{
+		0x000010025D071100,
+		1,
+		&single_endpoint,
+	}
```
Comment (lyakh) [inline sound/soc/intel/common/soc-acpi-intel-cml-match.c:88]: named?

File: sound/soc/intel/common/soc-acpi-intel-cml-match.c [hunk 4]
```diff
@@ -59,42 +59,96 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_cml_machines[] = {
...
+
```
Comment (lyakh) [inline sound/soc/intel/common/soc-acpi-intel-cml-match.c]: superfluous empty line

File: sound/soc/intel/common/soc-acpi-intel-icl-match.c [hunk 5]
```diff
@@ -33,55 +33,113 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_icl_machines[] = {
...
+	1, /* group_position */
+	1, /* group_id */
+};
+
+static const struct snd_soc_acpi_adr_device rt700_0_adr[] = {
+	{
+		0x000010025D070000,
+		1,
+		&single_endpoint,
+	}
```
Comment (lyakh) [inline sound/soc/intel/common/soc-acpi-intel-icl-match.c:62]: named everywhere above and below

File: sound/soc/intel/common/soc-acpi-intel-icl-match.c [hunk 6]
```diff
@@ -33,55 +33,113 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_icl_machines[] = {
...
+	{
+		0x000110025D130800,
+		1,
+		&spk_l_endpoint,
+	}
 };
 
-static const u64 rt715_3_adr[] = {
-	0x000310025D071500
+static const struct snd_soc_acpi_adr_device rt1308_2_group1_adr[] = {
```
Comment (bardliao) [inline sound/soc/intel/common/soc-acpi-intel-icl-match.c:106]: looks like rt1308_1_group1_adr[] and rt1308_2_group1_adr[] are never used in soc-acpi-intel-icl-match.c?

File: sound/soc/intel/common/soc-acpi-intel-icl-match.c [hunk 7]
```diff
@@ -33,55 +33,113 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_icl_machines[] = {
...
 		.mask = BIT(1),
 		.num_adr = ARRAY_SIZE(rt1308_1_adr),
-		.adr = rt1308_1_adr,
+		.adr_d = rt1308_1_adr,
 	},
 	{
 		.mask = BIT(2),
 		.num_adr = ARRAY_SIZE(rt1308_2_adr),
-		.adr = rt1308_2_adr,
+		.adr_d = rt1308_2_adr,
```
Comment (bardliao) [inline sound/soc/intel/common/soc-acpi-intel-icl-match.c:136]: And this should be rt1308_2_group1_adr?

File: sound/soc/intel/common/soc-acpi-intel-icl-match.c [hunk 8]
```diff
@@ -33,55 +33,113 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_icl_machines[] = {
...
 		.mask = BIT(0),
 		.num_adr = ARRAY_SIZE(rt711_0_adr),
-		.adr = rt711_0_adr,
+		.adr_d = rt711_0_adr,
 	},
 	{
 		.mask = BIT(1),
 		.num_adr = ARRAY_SIZE(rt1308_1_adr),
-		.adr = rt1308_1_adr,
+		.adr_d = rt1308_1_adr,
```
Comment (bardliao) [inline sound/soc/intel/common/soc-acpi-intel-icl-match.c:131]: I guess this should be rt1308_1_group1_adr?

File: sound/soc/intel/common/soc-acpi-intel-cml-match.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-cml-match.c]: ack

File: sound/soc/intel/common/soc-acpi-intel-cml-match.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-cml-match.c]: ack

File: sound/soc/intel/common/soc-acpi-intel-cml-match.c (hunk 3 above)
Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-cml-match.c:88]: ack

File: sound/soc/intel/common/soc-acpi-intel-cml-match.c (hunk 4 above)
Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-cml-match.c]: ack

File: sound/soc/intel/common/soc-acpi-intel-icl-match.c (hunk 5 above)
Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-icl-match.c:62]: ack

File: sound/soc/intel/common/soc-acpi-intel-icl-match.c (hunk 6 above)
Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-icl-match.c:106]: yes, fixed

File: sound/soc/intel/common/soc-acpi-intel-icl-match.c (hunk 7 above)
Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-icl-match.c:136]: yep, fixed

File: sound/soc/intel/common/soc-acpi-intel-icl-match.c (hunk 8 above)
Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-icl-match.c:131]: yep

**Comment by plbossart** [comment]
Thanks @bardliao and @lyakh for the review, all comments addressed

File: sound/soc/intel/common/soc-acpi-intel-cml-match.c [hunk 9]
```diff
@@ -59,42 +59,95 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_cml_machines[] = {
 };
 EXPORT_SYMBOL_GPL(snd_soc_acpi_intel_cml_machines);
 
-static const u64 rt711_0_adr[] = {
-	0x000010025D071100
+static const struct snd_soc_acpi_endpoint single_endpoint = {
+	.num = 0,
+	.aggregated = 0,
+	.group_position = 0,
+	.group_id = 0,
```
Comment (ranj063) [inline sound/soc/intel/common/soc-acpi-intel-cml-match.c:66]: @plbossart do you think it makes sense to have a macro for this?

File: include/sound/soc-acpi.h [hunk 10]
```diff
@@ -75,18 +75,47 @@ struct snd_soc_acpi_mach_params {
...
+ * snd_soc_acpi_adr_device: descriptor for device, include _ADR
+ * reresented at 64 bits and endpoint information
+ *
+ * @adr: 64 bit ACPI _ADR value
+ * @num_endpoints: number of endpoints for this device
+ * @endpoints: array of endpoints
+ */
+struct snd_soc_acpi_adr_device {
+	const u64 adr;
+	const u8 num_endpoints;
```
Comment (ranj063) [inline include/sound/soc-acpi.h:99]: @plbossart no objection to the code but for my educatio: num_endpoints doesnt seem to be more than 1 in any of the definitions below. Will it ever be?

Comment (plbossart) [reply include/sound/soc-acpi.h:99]: in most cases there will be a single endpoint per device, at least for the cases where we need to care about aggregation. that said it's probably better to plan for more in case someone combines two audio functions in the same device.
Note that the definition of endpoint is really minimal, e.g. for a headset codec I would most likely need two endpoints. I chose to keep things simple for now, this will be completely revisited anyways for new devices where the machine driver will have to be handled by parsing an ACPI description (similar to DT). The code in this PR is largely a placeholder until we have standardized info in the DSDT.

Comment (ranj063) [reply include/sound/soc-acpi.h:99]: Thanks @plbossart. I suppose I got carried away by the terminology and the definitions a bit. What confused me was the array of snd_soc_acpi_adr_device for each struct snd_soc_acpi_link_adr. It was an array of "u64 adr" even before. But I wonder why do we need an array here? Will there be more than one address per link?

Comment (plbossart) [reply include/sound/soc-acpi.h:99]: yes of course, on TGL RVP we have two RT1308 on the same link. In this case the concept of 'aggregation' is provided directly by SoundWire (same as TDM), so it's way simpler than when two links are used.

File: sound/soc/intel/common/soc-acpi-intel-cml-match.c (hunk 9 above)
Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-cml-match.c:66]: For now it's not really useful, we have only 3 definitions. We can always add macros later if needed.

File: include/sound/soc-acpi.h (hunk 10 above)
Comment (ranj063) [reply include/sound/soc-acpi.h:99]: ahh got it! Sorry I didnt look carefully enough to realize that. If you dont mind me asking one more dumb question: Who/what decides whether the 2 speaker amps go on the same link or different links? Is it purely based on board design?

Comment (plbossart) [reply include/sound/soc-acpi.h:99]: If you put two speakers on the same link, you increase the capacitive load and make the routing more complicated (all devices need to see the same clocks, otherwise you could have electrical issues).
If you put two speakers on different links, you can potentially simplify the routing with different wires going in different directions, BUT you will now need to rely on a synchronization of the two links with non-standard means (we have a 'gsync' hw signal on Intel platforms but that's not part of the MIPI standard). You also need to use 2 additional pins per link.
Also with more than one link, you can potentially make power management simpler, e.g. you only enable the links that are active, and go in clock-stop mode if not.
In short, the decision is going to be a compromise between hardware/software/routing/pin count considerations. I think for 2 speakers it's likely the same complexity overall. If you start adding more devices then at some point electrical limitations and bandwidth restrictions will force the use of more than one link.
Does this help?

Comment (ranj063) [reply include/sound/soc-acpi.h:99]: Yes, thank you very much!

File: include/sound/soc-acpi.h [hunk 11]
```diff
@@ -75,18 +75,47 @@ struct snd_soc_acpi_mach_params {
...
+	const struct snd_soc_acpi_endpoint *endpoints;
+};
+
+/**
+ * snd_soc_acpi_link_adr: ACPI-based list of devices identified with an
+ * _ADR, with a variable number of devices per link
```
Comment (lyakh) [inline include/sound/soc-acpi.h]: I'm not sure about this: I think usually the "short description" should be one line, does it still work with more than that? You can try to generate some documentation from this to check. Or just go with 1 line for safety and put any details below under all the struct members. Also I think this first line usually has the format of " * name - description" - not with a semicolon, but I'm not sure how hard that rule is either.

**Comment by RanderWang** [comment]
@plbossart   I implemented a new solution based on this PR. It is much simple than my last one. I will create a PR when this PR is merged.

File: include/sound/soc-acpi.h (hunk 11 above)
Comment (plbossart) [reply include/sound/soc-acpi.h]: will check, thanks for spotting this. I ran make W=1 which checks the kernel doc and didn't see any warning, but it doesn't hurt to check the documentation itself

Comment (plbossart) [reply include/sound/soc-acpi.h]: fixed with the proper format, thanks @lyakh
